### PR TITLE
WIP, MAINT: draft backports for 1.5.3

### PR DESCRIPTION
**Pull request by tylerjereddy**
- draft backports for SciPy `1.5.3`; not yet including shims for ARM64 support/CI testing
- unfortunately, merge conflicts were plentiful and just looking at the diff I probably need to do some more clean-up on that
- pretty much impossible that CI will pass on first try--we've had tons of tiny fixes here and there to the CI to keep things green in `master`, but let's see what the situation is at least

### CODE REVIEW

File: THANKS.txt [hunk 1]
```diff
@@ -239,6 +239,9 @@ Richard Weiss for a bug fix in scipy.optimize._differentialevolution.py.
 Luigi F. Cruz for adding time/frequency domain option to scipy.signal.resample.
 Wesley Alves for improvements to scipy.stats.jarque_bera and scipy.stats.shapiro
 Mark Borgerding for contributing linalg.convolution_matrix.
+Shashaank N for contributions to scipy.signal.
+Frank Torres for fixing a bug with solve_bvp for large problems.
+Ben West for updating the Gamma distribution documentation.
```
Comment (tylerjereddy) [inline THANKS.txt:244]: Remove these 3 lines

File: scipy/interpolate/interpnd.pyx [hunk 2]
```diff
@@ -257,8 +257,8 @@ class LinearNDInterpolator(NDInterpolatorBase):
     def _do_evaluate(self, double[:,::1] xi, double_or_complex dummy):
         cdef double_or_complex[:,::1] values = self.values
         cdef double_or_complex[:,::1] out
-        cdef double[:,::1] points = self.points
-        cdef int[:,::1] simplices = self.tri.simplices
+        cdef const double[:,::1] points = self.points
```
Comment (tylerjereddy) [inline scipy/interpolate/interpnd.pyx:260]: this first `const` not added in more conservative merge, but maybe ok

File: scipy/interpolate/interpnd.pyx [hunk 3]
```diff
@@ -854,12 +854,12 @@ class CloughTocher2DInterpolator(NDInterpolatorBase):
...
-        cdef double_or_complex[:,::1] values = self.values
-        cdef double_or_complex[:,:,:] grad = self.grad
+    def _do_evaluate(self, const double[:,::1] xi, double_or_complex dummy):
+        cdef const double_or_complex[:,::1] values = self.values
+        cdef const double_or_complex[:,:,:] grad = self.grad
         cdef double_or_complex[:,::1] out
-        cdef double[:,::1] points = self.points
-        cdef int[:,::1] simplices = self.tri.simplices
+        cdef const double[:,::1] points = self.points
+        cdef const int[:,::1] simplices = self.tri.simplices
```
Comment (tylerjereddy) [inline scipy/interpolate/interpnd.pyx:862]: less `const` changes in more conservative merge conflict resolution, but maybe "ok" (perhaps both sets of `const` deserved to be backported anyway..)

File: scipy/ndimage/tests/test_measurements.py [hunk 4]
```diff
@@ -1084,3 +1098,214 @@ def test_stat_funcs_2d():
     max = ndimage.maximum(a, labels=lbl, index=[1, 2])
     assert_array_equal(max, [9, 5])
 
+
+class TestWatershedIft:
```
Comment (tylerjereddy) [inline scipy/ndimage/tests/test_measurements.py:1102]: this entire test class snuck in with the backport, but really just the last test `test_watershed_ift08` was explicitly added in a backport PR; the more conservative test merge I did this evening just split that test off, but not sure if other tests are "harmful"

File: scipy/signal/filter_design.py [hunk 5]
```diff
@@ -4843,6 +4845,200 @@ def _design_notch_peak_filter(w0, Q, ftype, fs=2.0):
     return b, a
 
 
+def _hz_to_erb(hz):
```
Comment (tylerjereddy) [inline scipy/signal/filter_design.py:4848]: this entire block down to line `5040` was excluded in my more conservative test merge--not sure if the private function is actually harmful, but just noting that difference

**Review by tylerjereddy — COMMENTED** [review]
I did a more careful & conservative "test merge" of the same backport PRs this evening since the original set of merge conflict resolutions was a bit rushed for my taste. I've added inline review comments to highlight some areas where my decisions may benefit from some correction/verification.

**Comment by tylerjereddy** [comment]
There are 3 categories of CI failures left here at the moment, all confined to Travis CI. Since I've now backported in everything that was both merged and originally targeted for `1.6.0` with the backport label, these are presumably a bit less trivial to fix:

1. `pycodestyle scipy` fails with something like this `scipy/signal/filter_design.py:5042:1: E305 expected 2 blank lines after class or function definition`
2. with NumPy pre-wheels, a large number of errors that look like this: `ckdtree.pyx:917: in scipy.spatial.ckdtree.cKDTree.query_ball_point TypeError: 'numpy.float64' object cannot be interpreted as an integer`
3. For an `sdist` entry: `_csparsetools.pyx:1: in init scipy.sparse._csparsetools  RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject `

After those are fixed the approximate planned steps would include:
- fixing up some of the backport diff comments above if desired
- backporting in what is needed/appropriate for ARM64 CI testing & support (and wheels release)
- Python 3.9 may be delayed to `1.5.4` rather than squeezing in even more stuff here, until that ecosystem builds out a bit more (?)

**Comment by tylerjereddy** [comment]
@hameerabbasi @stuartarchibald 

For one of the failures here ( example log: https://travis-ci.org/github/scipy/scipy/jobs/733853108 ) that has errors with ` RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject`, I also see this much earlier in the log: `numba 0.49.1 requires numpy>=1.15, but you'll have numpy 1.14.5 which is incompatible`.

I'm not sure if the latter warning is just a red herring, but they seem connected at first glance, and we have had to carefully pin on matters related to `pydata/sparse` numba version reqs like in https://github.com/scipy/scipy/pull/11511

**Comment by tylerjereddy** [comment]
I may have to dissect the history on the changing of that `numba` version pin a bit; the line was reformatted in the most recent `git blame`, but version change happened before that. Probably wont' get to it until tomorrow though.

**Comment by stuartarchibald** [comment]
> @hameerabbasi @stuartarchibald
> 
> For one of the failures here ( example log: https://travis-ci.org/github/scipy/scipy/jobs/733853108 ) that has errors with ` RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject`, I also see this much earlier in the log: `numba 0.49.1 requires numpy>=1.15, but you'll have numpy 1.14.5 which is incompatible`.
> 
> I'm not sure if the latter warning is just a red herring, but they seem connected at first glance, and we have had to carefully pin on matters related to `pydata/sparse` numba version reqs like in #11511

Numba raised it's minimum supported versions earlier in the year https://github.com/numba/numba/blob/release0.51/CHANGE_LOG#L517-L519, and will refuse to work with anything < 1.15 https://github.com/numba/numba/blob/release0.51/numba/__init__.py#L131-L132

**Comment by rgommers** [comment]
I think using the `numba==0.47` version pin in CI for the `1.5.x` branch is fine. Release-wise there's nothing to do, because Numba is only an indirect optional dependency; it's up to pydata/sparse to set the correct lower and upper bounds.

Also thinking: yet another data point for using upper bounds on versions.

**Comment by tylerjereddy** [comment]
Everything passed except for that `numba`/`sparse` CI job. Pinning to an older version of `numba` actually results in a newer version of `numba` (`0.51.2`) getting dragged in via `sparse` it seems.

I don't know if this is the actual problem that Travis CI entry is having, but I'm now trying to additional pin `sparse` to a version (`0.9.1`) that has a `requirements.txt` that does not require a version of `numba>0.47`, so I had to go back several versions based on my git history checking. Anyway, fingers crossed..

**Comment by tylerjereddy** [comment]
Looks like that wasn't the real problem. NumPy version gaps between build time and runtime may vary between 1.14.5 and 1.17.4. Perhaps related to that; not sure if we can or want to suppress the warnings to avoid erroring out--I believe some measure of forward compatibility is in place for struct size changes.

**Comment by rgommers** [comment]
The `test_arpack` failure can simply be ignored here, that's a small test tolerance thing.

For the warning, let's just add a new commit on this branch with:
```
    # Filter out Cython harmless warnings
    warnings.filterwarnings("ignore", message="numpy.dtype size changed")
    warnings.filterwarnings("ignore", message="numpy.ufunc size changed")
```
in `scipy/__init__.py`. Those warnings got more persistent again somehow, which is annoying.

**Comment by tylerjereddy** [comment]
Argh, still didn't reproduce with a local `sdist`/`pip install` build cycle using this feature branch locally, but maybe the true versions used are deceptive or something.

**Comment by rgommers** [comment]
If it gets too painful, I suggest to (a) leave the new `filterwarnings` in in `scipy/__init__.py` because that can't hurt, and (b) turn off the "convert warnings to errors" so the test suite runs to completion. If that makes the CI green, let's ignore the issue.

**Comment by tylerjereddy** [comment]
Good idea, I went for the `pytest.ini` itself--let's hope that brute force change finally does the trick.

**Comment by tylerjereddy** [comment]
CI is all green now--my plan is to merge this & then proceed with a separate backport PR that pulls in the testing and shims/patches (if any) needed to support Linux ARM64.

**Comment by tylerjereddy** [comment]
@odidev @janaknat Can you just confirm for me that you don't know of any other ARM64-related shims that need to be backported to the `1.5.x` branch? I hadn't actually realized that we are actually running the ARM tests in the CI here already.

I believe it is just for 1 Python version and only the "fast" suite, so we may have a few surprises when I backport the wheel builds in the wheels repo I suppose. But I guess we can deal with that if/when it causes issues.

**Comment by janaknat** [comment]
@tylerjereddy The only commit I remember is the sha256 PR from @odidev that was needed to build aarch64 wheels.

PR: https://github.com/scipy/scipy/pull/12705

**Pull request merged**